src/pdn/src/grid.cpp

```diff
@@ -0,0 +1,1114 @@
+    }
+
+    if (box->getTechLayer() == nullptr) {
+      for (auto* layer : block->getDb()->getTech()->getLayers()) {
+        auto shape = std::make_shared<Shape>(layer, obs_rect, Shape::BLOCK_OBS);
+        obs[shape->getLayer()].insert({shape->getObstructionBox(), shape});
+      }
+    } else {
+      auto shape = std::make_shared<Shape>(
+          box->getTechLayer(), obs_rect, Shape::BLOCK_OBS);
+
+      obs[shape->getLayer()].insert({shape->getObstructionBox(), shape});
+    }
+  }
+
+  // placed block obs
+  for (auto* inst : block->getInsts()) {
+    if (!inst->isBlock()) {
+      continue;
+    }
+
+    if (!inst->isPlaced()) {
+      continue;
+    }
+
+    for (const auto& [layer, shapes] : InstanceGrid::getInstanceObstructions(inst)) {
+      obs[layer].insert(shapes.begin(), shapes.end());
+    }
+  }
+}
+
+void Grid::makeInitialShapes(const std::set<odb::dbNet*>& nets,
+                             ShapeTreeMap& shapes)
+{
+  for (auto* net : nets) {
+    Shape::populateMapFromDb(net, shapes);
+  }
+}
+
+std::set<odb::dbTechLayer*> Grid::connectableLayers(
+    odb::dbTechLayer* layer) const
+{
+  std::set<odb::dbTechLayer*> layers;
+
+  for (const auto& connect : connect_) {
+    if (connect->getLowerLayer() == layer) {
+      layers.insert(connect->getUpperLayer());
+    } else if (connect->getUpperLayer() == layer) {
+      layers.insert(connect->getLowerLayer());
+    }
+  }
+  return layers;
+}
+
+///////////////
+
+CoreGrid::CoreGrid(VoltageDomain* domain,
+                   const std::string& name,
+                   bool start_with_power)
+    : Grid(domain, name, start_with_power)
+{
+}
+
+const odb::Rect CoreGrid::getDomainBoundary() const
+{
+  // account for the width of the follow pins for straps
+  odb::Rect core = Grid::getDomainBoundary();
+
+  int follow_pin_width = 0;
+  for (const auto& strap : getStraps()) {
+    if (strap->type() == GridComponent::Followpin) {
+      follow_pin_width = std::max(follow_pin_width, strap->getWidth());
+    }
+  }
+
+  core.bloat(follow_pin_width / 2, core);
+  return core;
+}
+
+void CoreGrid::setupDirectConnect(
+    const std::vector<odb::dbTechLayer*>& connect_pad_layers)
+{
+  // look for pads that need to be connected
+  for (auto* net : getNets()) {
+    std::vector<odb::dbITerm*> iterms;
+    for (auto* iterm : net->getITerms()) {
+      auto* inst = iterm->getInst();
+      if (!inst->getPlacementStatus().isPlaced()) {
+        continue;
+      }
+
+      if (!inst->getMaster()->isPad()) {
+        continue;
+      }
+
+      iterms.push_back(iterm);
+    }
+
+    // sort by name to keep stable
+    std::stable_sort(
+        iterms.begin(), iterms.end(), [](odb::dbITerm* l, odb::dbITerm* r) {
+          const std::string r_name
+              = l->getInst()->getName() + "/" + l->getMTerm()->getName();
+          const std::string l_name
+              = l->getInst()->getName() + "/" + l->getMTerm()->getName();
+          return r_name < l_name;
```

Comment:
You could first compare the inst names and then the mterms names if needed rather the construct the full string.  string::compare is good for this.